### PR TITLE
Typecheck forbidden p3

### PR DESCRIPTION
**Pull request by smohsinali**

### CODE REVIEW

File: ConfigSpace/forbidden.py [hunk 1]
```diff
@@ -29,14 +29,16 @@
...
 
-from ConfigSpace.hyperparameters import Hyperparameter
 
+from ConfigSpace.hyperparameters import Hyperparameter
+from typing import List, Dict, Any, Union
 
 class AbstractForbiddenComponent(object):
     __metaclass__ = ABCMeta
-
+    # todo : should absract methods be annotated with types?
```
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: From my point of view, this is not necessary here. Did mypy complain? If yes, we have to do something about it. Else, please remove the comment. The other abstract methods which should return something should be annotated though.

File: ConfigSpace/forbidden.py [hunk 2]
```diff
@@ -72,12 +74,13 @@ def is_forbidden(self, instantiated_hyperparameters):
 
 
 class AbstractForbiddenClause(AbstractForbiddenComponent):
-    def get_descendant_literal_clauses(self):
+    def get_descendant_literal_clauses(self) -> List['AbstractForbiddenClause']:
```
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: Wouldn't `AbstractForbiddenComponent` be more appropriate her? Also, why is this a string? Moreover, the definition further below suggests that this function return `List[Hyperparameter].`

File: ConfigSpace/forbidden.py [hunk 3]
```diff
@@ -72,12 +74,13 @@ def is_forbidden(self, instantiated_hyperparameters):
...
         return [self]
 
 
 class SingleValueForbiddenClause(AbstractForbiddenClause):
-    def __init__(self, hyperparameter, value):
+    # todo: possible types of hyperparameter valuees?
```
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: I'm afraid we need to accept `Any` here. Please remove the comment.

File: ConfigSpace/forbidden.py [hunk 4]
```diff
@@ -89,7 +92,7 @@ def __init__(self, hyperparameter, value):
                              "'%s'" % (hyperparameter, str(value)))
         self.value = value
 
-    def is_forbidden(self, instantiated_hyperparameters, strict=True):
+    def is_forbidden(self, instantiated_hyperparameters: Hyperparameter, strict: bool=True) -> bool:
```
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: Please check the code by running it, but I think that `instantiated_hyperparameters` is a `dict`. Please also annotate `_is_forbidden`.

File: ConfigSpace/forbidden.py [hunk 5]
```diff
@@ -109,7 +112,8 @@ def _is_forbidden(self, target_instantiated_hyperparameter):
 
 
 class MultipleValueForbiddenClause(AbstractForbiddenClause):
-    def __init__(self, hyperparameter, values):
+    # todo: type of vals of hp
```
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: Same as above.

File: ConfigSpace/forbidden.py [hunk 6]
```diff
@@ -109,7 +112,8 @@ def _is_forbidden(self, target_instantiated_hyperparameter):
...
+    def __init__(self, hyperparameter: Hyperparameter, values: Any) -> None:
```
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: Same as above.

File: ConfigSpace/forbidden.py [hunk 7]
```diff
@@ -182,7 +186,7 @@ def __init__(self, *args):
     def __repr__(self):
         pass
 
-    def get_descendant_literal_clauses(self):
+    def get_descendant_literal_clauses(self) -> List[Hyperparameter]:
```
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: Please update this based on the comment above.

File: ConfigSpace/forbidden.py [hunk 8]
```diff
@@ -191,7 +195,7 @@ def get_descendant_literal_clauses(self):
                 children.append(component)
         return children
 
-    def is_forbidden(self, instantiated_hyperparameters, strict=True):
+    def is_forbidden(self, instantiated_hyperparameters: Hyperparameter, strict: bool=True) -> bool:
```
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: Same as above.

File: ConfigSpace/forbidden.py [hunk 9]
```diff
@@ -210,8 +214,7 @@ def is_forbidden(self, instantiated_hyperparameters, strict=True):
         # outcomes
         evaluations = []
         for component in self.components:
-            e = component.is_forbidden(instantiated_hyperparameters,
-                                       strict=strict)
+            e = component.is_forbidden(instantiated_hyperparameters)
```
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: Why this?

File: ConfigSpace/forbidden.py (hunk 1 above)
Comment (smohsinali) [inline ConfigSpace/forbidden.py]: no mypy didnt complain...removed this todo

File: ConfigSpace/forbidden.py (hunk 2 above)
Comment (smohsinali) [inline ConfigSpace/forbidden.py]: changed. 
also its different class so no need to declare it as str.
why string? http://stackoverflow.com/questions/33533148/how-do-i-specify-that-the-return-type-of-a-method-is-the-same-as-the-class-itsel

File: ConfigSpace/forbidden.py (hunk 3 above)
Comment (smohsinali) [inline ConfigSpace/forbidden.py]: removed

File: ConfigSpace/forbidden.py (hunk 4 above)
Comment (smohsinali) [inline ConfigSpace/forbidden.py]: changed the type of instantiated_hyperparameters to Dict[str, Union[None, str, float, int]] .
all versions of _is_forbidden are annotated except abstract methods

File: ConfigSpace/forbidden.py (hunk 5 above)
Comment (smohsinali) [inline ConfigSpace/forbidden.py]: done

File: ConfigSpace/forbidden.py (hunk 8 above)
Comment (smohsinali) [inline ConfigSpace/forbidden.py]: done

File: ConfigSpace/forbidden.py (hunk 7 above)
Comment (smohsinali) [inline ConfigSpace/forbidden.py]: AbstractForbiddenComponent or AbstractForbiddenConjunction?

File: ConfigSpace/forbidden.py (hunk 9 above)
Comment (smohsinali) [inline ConfigSpace/forbidden.py]: type of strict is bool so its value should be 'true' or 'false' and not 'strict'

File: ConfigSpace/forbidden.py (hunk 2 above)
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: Could you please check with the debugger and the unittests whether this actually returns a list of `AbstractForbiddenClause`? I would assume that the list can contain `AbstractForbiddenComponent` as well.

File: ConfigSpace/forbidden.py [hunk 10]
```diff
@@ -182,8 +182,8 @@ def __init__(self, *args):
     @abstractmethod
     def __repr__(self):
         pass
-
-    def get_descendant_literal_clauses(self):
+    # todo: recheck is return type should be AbstractForbiddenComponent or AbstractForbiddenConjunction or Hyperparameter
```
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: Could you please check this with the debugger?

File: ConfigSpace/forbidden.py (hunk 9 above)
Comment (mfeurer) [inline ConfigSpace/forbidden.py]: Yes, but strict is a boolean variable that is passed to `is_forbidden` in line 195; so it won't change till down here.

Comment (smohsinali) [inline ConfigSpace/forbidden.py]: fixed

File: ConfigSpace/forbidden.py (hunk 10 above)
Comment (smohsinali) [inline ConfigSpace/forbidden.py]: mypy accepts only List[hyperparameter] but according to debugger AbstractForbiddenComponent should also work

Comment (mfeurer) [inline ConfigSpace/forbidden.py]: What do you mean by also? Can both happen?

Comment (smohsinali) [inline ConfigSpace/forbidden.py]: from output of debugger it seems that the object of type forbidden has hyperparameter as its keys

Comment (smohsinali) [inline ConfigSpace/forbidden.py]: fixed by adding hyperparameter attribute to AbstractForbiddenComponent

**Pull request merged**